internal/controller/backup/collect.go

```diff
@@ -0,0 +1,137 @@
+	}
+
+	orSelectors := []*metav1.LabelSelector{
+		// fixed ones
+		selector(kcmv1alpha1.GenericComponentNameLabel, kcmv1alpha1.GenericComponentLabelValueKCM),
+		selector(certmanagerv1.PartOfCertManagerControllerLabelKey, "true"),
+		selector(clusterapiv1beta1.ProviderNameLabel, "cluster-api"),
+	}
+
+	clusterTemplates := new(kcmv1alpha1.ClusterTemplateList)
+	if err := c.cl.List(ctx, clusterTemplates); err != nil {
+		return nil, fmt.Errorf("failed to list ClusterTemplates: %w", err)
+	}
+
+	if len(clusterTemplates.Items) == 0 { // just collect child clusters names
+		cldSelectors, err := getClusterDeploymentsSelectors(ctx, c.cl, "")
+		if err != nil {
+			return nil, fmt.Errorf("failed to get selectors for all clusterdeployments: %w", err)
+		}
+
+		bs.OrLabelSelectors = sortDedup(append(orSelectors, cldSelectors...))
+
+		return bs, nil
+	}
+
+	for _, cltpl := range clusterTemplates.Items {
+		cldSelectors, err := getClusterDeploymentsSelectors(ctx, c.cl, cltpl.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get selectors for clusterdeployments referencing %s clustertemplate: %w", client.ObjectKeyFromObject(&cltpl), err)
+		}
+
+		// add only enabled providers
+		if len(cldSelectors) > 0 {
+			for _, provider := range cltpl.Status.Providers {
+				orSelectors = append(orSelectors, selector(clusterapiv1beta1.ProviderNameLabel, provider))
+			}
+		}
+
+		orSelectors = append(orSelectors, cldSelectors...)
+	}
+
+	bs.OrLabelSelectors = sortDedup(orSelectors)
+
+	return bs, nil
+}
+
+func sortDedup(selectors []*metav1.LabelSelector) []*metav1.LabelSelector {
+	const nonKubeSep = "_"
+
+	kvs := make([]string, len(selectors))
+	for i, s := range selectors {
+		for k, v := range s.MatchLabels { // expect only one kv pair
+			kvs[i] = k + nonKubeSep + v
+		}
+	}
+	slices.Sort(kvs)
+
+	for i, kv := range kvs {
+		sepIdx := strings.Index(kv, nonKubeSep)
+		if sepIdx < 0 {
+			continue // make compiler happy
+		}
+		k := kv[:sepIdx]
+		v := kv[sepIdx+len(nonKubeSep):]
+		selectors[i] = selector(k, v)
+	}
+
+	return slices.Clip(
+		slices.CompactFunc(selectors, func(a, b *metav1.LabelSelector) bool {
+			return maps.Equal(a.MatchLabels, b.MatchLabels)
+		}),
+	)
+}
+
+func getClusterDeploymentsSelectors(ctx context.Context, cl client.Client, clusterTemplateRef string) ([]*metav1.LabelSelector, error) {
+	cldeploys := new(kcmv1alpha1.ClusterDeploymentList)
+	opts := []client.ListOption{}
+	if clusterTemplateRef != "" {
+		opts = append(opts, client.MatchingFields{kcmv1alpha1.ClusterDeploymentTemplateIndexKey: clusterTemplateRef})
+	}
+
+	if err := cl.List(ctx, cldeploys, opts...); err != nil {
+		return nil, fmt.Errorf("failed to list ClusterDeployments: %w", err)
+	}
+
+	selectors := make([]*metav1.LabelSelector, len(cldeploys.Items)*2)
+	for i, cldeploy := range cldeploys.Items {
+		selectors[i] = selector(kcmv1alpha1.FluxHelmChartNameKey, cldeploy.Name)
+		selectors[i+1] = selector(clusterapiv1beta1.ClusterNameLabel, cldeploy.Name)
```

Comment:
The `i` variable corresponds to the index of `cldeploys.Items` but when you assign `selectors[i]` this value may already be assigned on the previous iteration by `selectors[i+1] = ...`. Is it correct?

Reply:
right, thanks for noting, fixed!